### PR TITLE
Update to zig master, move uf2 and regz to monorepo

### PR DESCRIPTION
**Pull request by mattnite**

### CODE REVIEW

**Comment by MasterQ32** [comment]
Before merging, we have to branch into a zig-0.11.0 branch which i will backport as long as feasible

**Comment by MasterQ32** [comment]
Also can we incorporate the `microzig-package.json` into `build.zig.zon` for the packages? Afaik Zig will ignore additional fields, so we can encode the data at a single point

**Comment by mattnite** [comment]
@MasterQ32 need some advice on how we want to deal with a breakage:

```
/Users/mattnite/zig/0.12.0-dev.3076+6e078883e/files/lib/std/std.zig:206:34: error: ambiguous reference
pub const options: Options = if (@hasDecl(root, "std_options")) root.std_options else .{};
                                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
core/src/start.zig:19:5: note: declared here
pub const std_options = if (@hasDecl(app, "std_options"))
~~~~^~~~~
examples/raspberrypi/rp2040/src/usb_hid.zig:150:5: note: declared here
pub const std_options = struct {
~~~~^~~~~
referenced by:
    scope_levels: /Users/mattnite/zig/0.12.0-dev.3076+6e078883e/files/lib/std/log.zig:115:25
    logEnabled: /Users/mattnite/zig/0.12.0-dev.3076+6e078883e/files/lib/std/log.zig:130:17
    log__anon_6179: /Users/mattnite/zig/0.12.0-dev.3076+6e078883e/files/lib/std/log.zig:123:19
    info__anon_3903: /Users/mattnite/zig/0.12.0-dev.3076+6e078883e/files/lib/std/log.zig:194:16
    task: core/src/core/usb.zig:95:40
    main: examples/raspberrypi/rp2040/src/usb_hid.zig:177:23
    microzig_main: core/src/start.zig:82:18
 ```

It looks like `root` simultaneously refers to the declarations of all modules in a compilation. We had a really slick system before where we'd override defaults that made sense for software made with MicroZig, but still allowed users to override as they pleased. All I can think of now is to set `std_options` in `start.zig`, and then we'll have to forward options from `microzig_options` to `std_options`.

**Comment by MasterQ32** [comment]
> It looks like `root` simultaneously refers to the declarations of all modules in a compilation. We had a really slick system before where we'd override defaults that made sense for software made with MicroZig, but still allowed users to override as they pleased.

I guess @andrewrk should take a look at this here. Sounds like a bug in Zig, but might also be by design.

> All I can think of now is to set `std_options` in `start.zig`, and then we'll have to forward options from `microzig_options` to `std_options`.

But i'd be fine with this.

File: build/build.zig.zon [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+            .path = "../core",
+        },
+    },
+
+    .paths = .{
+        "README.md",
+        "build.zig",
+        "build.zig.zon",
+        "LICENSE",
+    },
```
Comment (MasterQ32) [inline build/build.zig.zon:21]: This is missing the `src/generate_linkerscript.zig` file, `README.md` does not exist, `LICENCE` is missing.

I guess we should perform a final cleanup in the end, and also *only* package files present in `.paths` to the website?

File: examples/raspberrypi/rp2040/build.zig [hunk 2]
```diff
@@ -0,0 +1,62 @@
+const std = @import("std");
+const MicroZig = @import("microzig/build");
+const rp2040 = @import("microzig/bsp/raspberrypi/rp2040");
+
+const available_examples = [_]Example{
+    // RaspberryPi Boards:
+    .{ .target = rp2040.boards.raspberrypi.pico, .name = "pico_adc", .file = "src/adc.zig" },
```
Comment (MasterQ32) [inline examples/raspberrypi/rp2040/build.zig:7]: I'm fine with that change back to what we had before, but i'd add a generic function to add `findTarget` back somehow.
We don't have to expose it as the primary interface, but it makes things easier if we want to pass the target device from the outside

**Review by MasterQ32 — COMMENTED** [review]
Really quick review, didn't go through the big changes. Example usage on RP2040 looks fine so far, and this is the most important part of the whole project anyways

**Comment by mattnite** [comment]
Re: std_options changing to a struct instance: Actually I like this change, it was pretty easy to do the same for interrupts and it improves discoverability as zls completions help let the user know what interrupts they can define.

In addition, turning of building of examples for AVR for now. I see the following issue:

```
// error: the following command failed with 1 compilation errors:
//Users/mattnite/zig/0.12.0-dev.3097+5c0766b6c/files/zig build-exe -freference-trace=256 -OReleaseSmall -target avr-freestanding-eabi -mcpu avr5 --dep app --dep microzig -Mroot=/Users/mattnite/code/microzig/core/src/start.zig --dep microzig -Mapp=/Users/mattnite/code/microzig/examples/microchip/avr/src/blinky.zig --dep config --dep chip --dep cpu --dep hal --dep board -Mmicrozig=/Users/mattnite/code/microzig/core/src/microzig.zig -Mconfig=/Users/mattnite/code/microzig/zig-cache/c/303c67fccae4ec4bb03ec2180082b67b/options.zig --dep microzig -Mchip=/Users/mattnite/code/microzig/zig-cache/o/2c86b0de714a8b4409eb761d16297c00/chip.zig --dep microzig -Mcpu=/Users/mattnite/code/microzig/core/src/cpus/avr5.zig --dep microzig -Mhal=/Users/mattnite/code/microzig/bsp/microchip/avr/src/hals/ATmega328P.zig --dep microzig -Mboard=/Users/mattnite/code/microzig/bsp/microchip/avr/src/boards/arduino_uno.zig --cache-dir /Users/mattnite/code/microzig/zig-cache --global-cache-dir /Users/mattnite/.cache/zig --name arduino-nano_blinky -static -fcompiler-rt --script /Users/mattnite/code/microzig/zig-cache/o/2c186d936508aa71bea517796451c3f9/linker.ld --listen=-
//install
//mq install
//   mq install generated to arduino-nano_blinky.hex
//      mq objcopy generated
//         mq zig build-exe arduino-nano_blinky ReleaseSmall avr-freestanding-eabi 1 errors
///Users/mattnite/zig/0.12.0-dev.3097+5c0766b6c/files/lib/std/debug.zig:403:14: error: reached unreachable code
//    if (!ok) unreachable; // assertion failure
//             ^~~~~~~~~~~
///Users/mattnite/zig/0.12.0-dev.3097+5c0766b6c/files/lib/std/os/wasi.zig:12:11: note: called from here
//    assert(@alignOf(i16) == 2);
//    ~~~~~~^~~~~~~~~~~~~~~~~~~~
```

**Comment by MasterQ32** [comment]
> In addition, turning of building of examples for AVR for now.

Yeah that might be reasonable. I think we should focus on the refactor, then start fixing the AVR stuff 

> I see the following issue …

That looks like someone assumed overaligned types :D Afaik, AVR allows everything to be `align(1)`

> Re: std_options changing to a struct instance: Actually I like this change, it was pretty easy to do the same for interrupts and it improves discoverability as zls completions help let the user know what interrupts they can define.

Hell, yeah!

**Comment by xdBronch** [comment]
in case this hasnt been said anywhere the alignment assertion problem is https://github.com/ziglang/zig/issues/19665
the std_options change accidentally made some wasi checks run too eagerly

**Pull request merged**